pkg/controller/pkg/manager/reconciler.go

```diff
@@ -0,0 +1,428 @@
+func SetupConfiguration(mgr ctrl.Manager, l logging.Logger) error {
+	name := "packages/" + strings.ToLower(v1alpha1.ConfigurationGroupKind)
+	np := func() v1alpha1.Package { return &v1alpha1.Configuration{} }
+	nr := func() v1alpha1.PackageRevision { return &v1alpha1.ConfigurationRevision{} }
+	nrl := func() v1alpha1.PackageRevisionList { return &v1alpha1.ConfigurationRevisionList{} }
+
+	r := NewReconciler(mgr,
+		WithInstallNamespace(os.Getenv(envPodNamespace)),
+		WithNewPackageFn(np),
+		WithNewPackageRevisionFn(nr),
+		WithNewPackageRevisionListFn(nrl),
+		WithLogger(l.WithValues("controller", name)),
+		WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor("configuration"))),
+	)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&v1alpha1.Configuration{}).
+		Owns(&v1alpha1.ConfigurationRevision{}).
+		Complete(r)
+}
+
+// NewReconciler creates a new package reconciler.
+func NewReconciler(mgr ctrl.Manager, opts ...ReconcilerOption) *Reconciler {
+	r := &Reconciler{
+		client: resource.ClientApplicator{
+			Client:     mgr.GetClient(),
+			Applicator: resource.NewAPIPatchingApplicator(mgr.GetClient()),
+		},
+		log:    logging.NewNopLogger(),
+		record: event.NewNopRecorder(),
+	}
+
+	for _, f := range opts {
+		f(r)
+	}
+
+	return r
+}
+
+// Reconcile package.
+func (r *Reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) { // nolint:gocyclo
+	log := r.log.WithValues("request", req)
+	log.Debug("Reconciling")
+
+	ctx, cancel := context.WithTimeout(context.Background(), reconcileTimeout)
+	defer cancel()
+
+	p := r.newPackage()
+	if err := r.client.Get(ctx, req.NamespacedName, p); err != nil {
+		// There's no need to requeue if we no longer exist. Otherwise we'll be
+		// requeued implicitly because we return an error.
+		log.Debug(errGetPackage, "error", err)
+		return reconcile.Result{}, errors.Wrap(resource.IgnoreNotFound(err), errGetPackage)
+	}
+
+	// Get existing package revisions.
+	prs := r.newPackageRevisionList()
+	if err := r.client.List(ctx, prs, client.MatchingLabels(map[string]string{parentLabel: p.GetName()})); err != nil {
+		log.Debug(errListRevisions, "error", err)
+		return reconcile.Result{}, errors.Wrap(resource.IgnoreNotFound(err), errListRevisions)
+	}
+
+	// Identify or create an install job.
+	// NOTE(hasheddan): the only advantage to using a job instead of just a pod
+	// is that the job will restart a failed pod automatically. However, if the
+	// job completes and the pod is then deleted, the job will not automatically
+	// recreate it. We handle this case below.
+	j := &batchv1.Job{}
+
+	// NOTE(hasheddan): we have to create the job to get the image ID of the
+	// package. If we move to parsing the image manifest from registry directly
+	// then we can instead create the package revision and let it manage the
+	// full lifecycle of its job.
+	jobName := strings.Replace(strings.Replace(p.GetSource(), "/", "-", -1), ":", "-", -1)
```

Comment:
This is a bit gross but gets the job done. In other places we have deferred to naming the "thing" associated with a package revision the hash of its image. However, we can't do that here because we don't know the hash yet. If we adjust the names of things like a Provider deployment or SA, would be good to follow that convention here.

Reply:
(Obviously only to the extent that we can without using the hash)

Reply:
@negz yep, but we would also collide at the `_____Revision` level, so likely okay to go ahead and hit that here.

Reply:
Would that be true even if a config and a package both tried to use the same image? It could be that that’s an obscure enough edge case that we shouldn’t bother with it.

Reply:
So there are a few cases here (I think no. 4 is what you are getting at):
1. Two providers use the same image name/tag with the same contents (hash). Right now they would collide on the `Job` and the second would be unable to install. If we made the `Job` unique, they would not collide, but would collides when the second tried to create the `ProviderRevision.
2. Two providers use the same image name/tag with different contents (hash). Right now they would collide on the `Job` and the second would be unable to install. If we made the `Job` unique, they would not collide and the `ProviderRevision` would not collide, making the second able to be installed as well (assuming no conflicting objects). However, this could be a somewhat confusing scenario and I am not sure we want to support it.
3. A provider and configuration use the same image name/tag with the same contents (hash). Right now they would collide on `Job`. If we made the `Job` unique, they would not collide until they got to installing objects. One of them would have to have a mismatch with the meta package type in the package, which would cause it to fail to install.
4. A provider and configuration use the same image name/tag with different contents (hash). Right now they would collide on `Job`. If we made `Job` unique, they would not collide, and they would not collide on `___Revision` or when installing objects (assuming the contents fundamentally changed the package type). This would technically be a valid use-case, but I once again am skeptical that we should support it. Using image hashes allow us to accommodate situations in which a package's contents change for the image with the same tag, and a user wants to update to using the new image contents (this could be done by deleting the existing `Job`). However, I don't think this is going to be a common use-case, and I don't think we should further complicate the model by allowing the use of the same image with different contents outside the scope of a single `Provider` or `Configuration`.

Reply:
If we go with the `providername-abbreviatedhash` naming scheme, will that take care of all of these cases?

Reply:
Yes 👍